lib/deimos/active_record_consume/batch_consumption.rb

```diff
@@ -86,38 +87,63 @@ def update_database(messages)
       # records to either be updated or inserted.
       # @return [void]
       def upsert_records(messages)
-        key_cols = key_columns(messages)
-
-        # Create payloads with payload + key attributes
-        upserts = messages.map do |m|
-          attrs = if self.method(:record_attributes).parameters.size == 2
-                    record_attributes(m.payload, m.key)
-                  else
-                    record_attributes(m.payload)
-                  end
-
-          attrs&.merge(record_key(m.key))
-        end
+        key_cols = key_columns(messages, nil)
 
+        # Create ActiveRecord Models with payload + key attributes
+        upserts = build_records(messages)
         # If overridden record_attributes indicated no record, skip
         upserts.compact!
+        # apply ActiveRecord validations and fetch valid Records
+        valid_upserts = filter_records(upserts)
+
+        return if valid_upserts.empty?
+
+        save_records_to_database(@klass, key_cols, valid_upserts)
+        import_associations(valid_upserts) unless @association_list.blank?
+      end
+
+      def save_records_to_database(record_class, key_cols, records)
+        columns = columns(record_class)
 
         options = if key_cols.empty?
                     {} # Can't upsert with no key, just do regular insert
-                  elsif ActiveRecord::Base.connection.adapter_name.downcase =~ /mysql/
+                  elsif mysql_adapter?
                     {
-                      on_duplicate_key_update: :all
+                      on_duplicate_key_update: columns
                     }
                   else
                     {
                       on_duplicate_key_update: {
                         conflict_target: key_cols,
-                        columns: :all
+                        columns: columns
                       }
                     }
                   end
+        record_class.import!(columns, records, options)
+      end
+
+      # Imports associated objects and import them to database table
+      # The base table is expected to contain bulk_import_id column for indexing associated objects with id
+      # @association_list configured on the consumer helps identify the ones required to be saved.
+      def import_associations(entities)
+        _validate_associations(entities)
+        _fill_primary_key_on_entities(entities)
+
+        # Select associations from config parameter association_list and
+        # fill id to associated_objects foreign_key column
+        @klass.reflect_on_all_associations.select { |assoc| @association_list.include?(assoc.name) }.
+          each do |assoc|
+            details = entities.map { |entity|
```

Comment:
Can we call this something like `sub_records` rather than `details`?

Reply:
Done.